runtime/doc/lua-async.txt

```diff
@@ -0,0 +1,630 @@
+      • {get_nowait}  (`fun(self: vim.async.Queue)`) See |Queue:get_nowait()|.
+      • {put_nowait}  (`fun(self: vim.async.Queue, value: any)`) See
+                      |Queue:put_nowait()|.
+
+*vim.async.Semaphore*
+    A semaphore manages an internal counter which is decremented by each
+    `acquire()` call and incremented by each `release()` call. The counter can
+    never go below zero; when `acquire()` finds that it is zero, it blocks,
+    waiting until some task calls `release()`.
+
+    The preferred way to use a Semaphore is with the `with()` method, which
+    automatically acquires and releases the semaphore around a function call.
+
+    Fields: ~
+      • {with}     (`fun(self: vim.async.Semaphore, fn: async fun(): R...): R...`)
+                   See |Semaphore:with()|.
+      • {acquire}  (`fun(self: vim.async.Semaphore)`) See
+                   |Semaphore:acquire()|.
+      • {release}  (`fun(self: vim.async.Semaphore)`) See
+                   |Semaphore:release()|.
+
+*vim.async.Task*
+    Tasks are used to run coroutines in event loops. If a coroutine needs to
+    wait on the event loop, the Task suspends the execution of the coroutine
+    and waits for event loop to restart it.
+
+    Use the |vim.async.run()| to create Tasks.
+
+    To cancel a running Task use the `close()` method. Calling it will cause
+    the Task to throw a "Task is closing or closed" error into the wrapped
+    coroutine.
+
+    Note a Task can be waited on via more than one waiter.
+
+    -- If a -- coroutine is awaiting on a Future object during cancellation,
+    the Future -- object will be cancelled.
+
+    Fields: ~
+      • {wait}            (`fun(self: vim.async.Task, callback_or_timeout: integer|fun(err?: any, ...: R...)?)`)
+                          See |Task:wait()|.
+      • {traceback}       (`fun(self: vim.async.Task, msg: string?): string`)
+                          See |Task:traceback()|.
+      • {raise_on_error}  (`fun(self: vim.async.Task): vim.async.Task`) See
+                          |Task:raise_on_error()|.
+      • {close}           (`fun(self: vim.async.Task, callback: fun(closed: boolean)?)`)
+                          See |Task:close()|.
+      • {status}          (`fun(self: vim.async.Task): 'running'|'suspended'|'normal'|'dead'?`)
+                          See |Task:status()|.
+
+*vim.vim.async.Future*
+    Future objects are used to bridge low-level callback-based code with
+    high-level async/await code.
+
+    Fields: ~
+      • {completed}  (`fun(self: vim.vim.async.Future): boolean`) See
+                     |Future:completed()|.
+      • {result}     (`fun(self: vim.vim.async.Future): boolean, any`) See
+                     |Future:result()|.
+      • {wait}       (`fun(self: vim.vim.async.Future, callback: fun(err?: any, ...: any))`)
+                     See |Future:wait()|.
+      • {complete}   (`fun(self: vim.vim.async.Future, err: string?, ...: any)`)
```

Comment:
I would expect a Task to have these properties too. Do we need both Future and Task, or can we centralized on Task ?

Reply:
`vim.async.Task` is implemented with `vim.async.Future`. `Future` is mostly just `Task` without all the coroutine stuff; a results holder which you can register callbacks on.

Can hide it from the API if you want, though out of everything, I consider the low level control primitives to be low risk. The high risk stuff is `run`, `await`, and the general concurrency model.

Reply:
> I consider the low level control primitives to be low risk

We could start by hiding it, and then expose the low-level things later. But still keep the low-level stuff clearly separated in docs/etc. Will be great for users if there is a clear "high level path" that they can follow, and they don't need to look at the low-level stuff unless/until they really care about that.